tests/end_to_end/test_op.py

```diff
@@ -97,3 +100,76 @@ def test_binops(execute):
     assert res.values["r10"] == 11
     assert res.values["r11"] == 9
     assert res.values["r12"] == 2
+
+
+@pytest.mark.skipif("sys.version_info < (3, 8)")
+def test_walrus_assigning_to_internal_identifier(execute):
+    code = """
+import lineapy
+x = (y := 10)
+(z := x) < 10
+(x := 7) > 9
+(a := z)
+(z := x) > 9
+lineapy.save(z, 'z')
+"""
+
+    res = execute(code, snapshot=False)
+    assert res.values["z"] == 7
+    assert res.artifacts["z"] == """(x := 7) > 9\n(z := x) > 9\n"""
+
+
+@pytest.mark.skipif("sys.version_info < (3, 8)")
+def test_walrus_assigning_using_returned_value(execute):
+    code = """
+z = 10
+z = (x := 8)
+"""
+    res = execute(code, snapshot=False)
+    assert res.values["z"] == 8
+
+
+@pytest.mark.skipif("sys.version_info < (3, 8)")
+def test_walrus_list_comprehensions(execute):
+    code = "(x := [t for t in range(3)])"
+
+    res = execute(code, snapshot=False)
+    assert res.values["x"] == [0, 1, 2]
+
+
+@pytest.mark.skipif("sys.version_info < (3, 8)")
+def test_walrus_multiple_identifiers(execute):
+    code = """
+import lineapy
+x = 1
+(x,y:=(1,2))
+lineapy.save(y, 'y')
+"""
+    res = execute(code, snapshot=False)
+    assert res.values["x"] == 1
+    assert res.values["y"] == (1, 2)
+    assert res.artifacts["y"] == "(x, y := (1, 2))\n"
```

Comment:
this looks wrong...

Reply:
i'm ok to let this get merged. we'll add xfail to denote that this syntax doesnt slice properly.